### PR TITLE
[feature] Enable net types to be used in hashed collections

### PR DESCRIPTION
**Pull request by hiteshjasani**
I wanted to use AsyncSockets in a set and ran into missing hash functions.

Some notes for reviewers:

* Only a subset of fields are being hashed on `AsyncSocket`.
  * The fields were chosen based on how much they represented the identity of the object.
  * The thinking is that if the `SocketHandle` or `Domain` were to change, then it is an entirely different object.  But if the `currPos` in the buffer were to change, then it's not important from an identity point of view.
* Are `enum`s a type of `Ordinal`?
  * Some documentation says they are and then some code says they are not.  For example, `system.ord` has a type signature that accepts either one.
  * `nativesockets.nim:199` failed to compile until `enum` was added to the type signature for `hash*[T: Ordinal](x: T)`.  The compiler didn't know how to generate a hash for the `Domain` enum.
  * Is there a better way to solve this problem?

### CODE REVIEW

**Comment by GULPF** [comment]
> Are enums a type of Ordinal?

An enum type will only match the `system.Ordinal` type if it has no holes (the `Domain` enum has a hole). Your change in `hashes` is correct, enums with holes should be hashable.

**Comment by krux02** [comment]
What GULPF said should be a comment in the code.

**Comment by Araq** [comment]
Ok, but every type that offers ``hash`` should also come with ``==``. I think.

File: lib/pure/asyncnet.nim [hunk 1]
```diff
@@ -130,6 +127,23 @@ type
...
+  ## Generates a hash using only a subset of the fields.  The ones
+  ## used are considered to provide identity.
+  var h: Hash = 0
+  h = h !& hash(x.fd)
+  h = h !& hash(x.domain)
+  h = h !& hash(x.sockType)
+  h = h !& hash(x.protocol)
+  result = !$h
+
+proc `==`*(a, b: AsyncSocket): bool =
```
Comment (Araq) [inline lib/pure/asyncnet.nim]: But AsyncSocket is a ``ref`` type, it's not intuitive if equality for AsyncSocket is overwritten. We need a better solution.

**Comment by Araq** [comment]
Sorry, but rejected. Equality for these types is inherently problematic.

**Pull request closed without merging**